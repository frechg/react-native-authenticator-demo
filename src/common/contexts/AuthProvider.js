import React from 'react'
import * as SecureStore from 'expo-secure-store';
import { initialAuthState, authStateReducer } from '../authStateReducer';
import * as Api from '../../services/auth';

export const AuthContext = React.createContext();

export const AuthProvider = (props) => {
  const [authState, authDispatch] = React.useReducer(authStateReducer, initialAuthState);

  const getAuthState = async () => {
    try {
      const token = await SecureStore.getItemAsync('authToken');
      const username = await SecureStore.getItemAsync('username');
      authDispatch({type: 'RESTORE_TOKEN', token: token, username: username});
    } catch (error) {
      throw new Error(error);
    }
  };

  const signIn = async (data) => {
    await Api.signIn(data)
    .then(async (response) => {
      if (response.ok) {
        const jsonResponse = await response.json();
        await SecureStore.setItemAsync('authToken', jsonResponse.token);
        await SecureStore.setItemAsync('username', jsonResponse.username);
        authDispatch({type: 'SIGN_IN', token: jsonResponse.token, username: jsonResponse.username});
      } else {
        throw new Error('Email or password are incorrect.');
      }
    })
  };

  const signUp = async (data) => {
    await Api.signUp(data)
    .then(async (response) => {
      if (response.ok) {
        const jsonResponse = await response.json();
        await SecureStore.setItemAsync('authToken', jsonResponse.token);
        await SecureStore.setItemAsync('username', jsonResponse.username);
        authDispatch({type: 'SIGN_IN', token: jsonResponse.token, username: jsonResponse.username});
      } else {
        const jsonResponse = await response.json();
        throw new Error(jsonResponse.errors[0])
      }
    });
  };

  const signOut = async () => {
    try {
      await SecureStore.deleteItemAsync('authToken');
      await SecureStore.deleteItemAsync('username');
      authDispatch({type: 'SIGN_OUT'});
    } catch (error) {
      throw new Error(error);
    }
  };

  const contextValue = React.useMemo(() => {
    return {authState, getAuthState, signUp, signIn, signOut};
  }, [authState]);

  return (
    <AuthContext.Provider value={contextValue}>
      {props.children}
    </AuthContext.Provider>
  );
};