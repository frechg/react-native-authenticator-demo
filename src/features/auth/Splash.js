import React from 'react';
import { ActivityIndicator, View } from 'react-native';
import { styles } from '../../common/styles';
import { AuthContext } from '../../common/contexts/AuthProvider';

export const Splash = () => {
  const { getAuthState } = React.useContext(AuthContext);

  React.useEffect(() => {
   getAuthState();
  }, []);

  return (
    <View style={styles.container}>
      <ActivityIndicator testID='LOADING' size='large' />
    </View>
  );
}