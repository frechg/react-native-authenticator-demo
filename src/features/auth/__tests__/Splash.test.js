import * as React from 'react';
import { render, screen } from '@testing-library/react-native';

import { Splash } from '../Splash';
import { AuthContext } from '../../../common/contexts/AuthProvider';

describe('<Splash />', () => {
  test('Renders Splash screen', () => {
    const getAuthState = jest.fn();
    render(
      <AuthContext.Provider value={{getAuthState}}>
        <Splash />
      </AuthContext.Provider>
    );

    expect(screen.toJSON()).toMatchSnapshot();
  });

  test('Checks current auth state', () => {
    const getAuthState = jest.fn();
    render(
      <AuthContext.Provider value={{getAuthState}}>
        <Splash />
      </AuthContext.Provider>
    );

    expect(getAuthState).toHaveBeenCalled();
  });
});
