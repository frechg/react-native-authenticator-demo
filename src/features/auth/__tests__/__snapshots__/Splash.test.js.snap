// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Splash /> Renders Splash screen 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    size="large"
    testID="LOADING"
  />
</View>
`;
